rfcs/0009-separate-wal.md

```diff
@@ -0,0 +1,99 @@
+# Separate Object Store for WAL
+
+Status: Draft
+
+Authors:
+
+* [Sergei Sitnikov](https://github.com/taburet)
+
+## Motivation
+
+Sequential writes are a common pattern. Combined with `await_durable` set to
+`true`, which is the default, WAL performance is the main limiting factor for
+the sequential writes throughput. Even if `flush_interval` is set to an extremely
+low value, it can't overcome the inherent latency of the underlying object store.
+
+The obvious solution is switching to a lower latency object store; for instance,
+from S3 to S3 Express One Zone. But lower latency stores usually come with a cost,
+they might be significantly more expensive or might be limited in features. Using
+such stores for the main store containing the entire database might be undesirable.
+
+## Goals
+
+- Support an optional separate object store dedicated specifically for WAL.
+- Provide a user-facing configuration mechanism.
+- Avoid significant refactorings.
+
+## Public API
+
+The only change on the public API level is configuration:
+
+```rust
+pub struct DbOptions {
+    // ...
+
+    /// An optional [object store](ObjectStore) dedicated specifically for WAL.
+    ///
+    /// If not set, the main object store passed to `Db::open(...)` will be used
+    /// for WAL storage.
+    ///
+    /// NOTE: WAL durability and availability properties depend on the properties
+    ///       of the underlying object store. Make sure the configured object
+    ///       store is durable and available enough for your use case.
```

Comment:
copying the comment from #504 

Adding few things I looked at.

https://aws.amazon.com/s3/pricing/?nc=sn&loc=4 - S3 pricing as of April 2025

Storage
S3 - 0.023 per GB
S3 express one - 0.16 per GB ( 7x more cost)

Put calls per 1000 requests
S3 - 0.005
S3 express one - 0.0025 (2x cheaper)
Get is 2x cheaper as well.

Keeping WAL in a different bucket could indeed be cheaper overall.

Shouldn't we extend it to all small files? WAL, L0, and manifest - all these would be small in size, but will have large number of operations.

Reply:
Responded here:

https://github.com/slatedb/slatedb/issues/504#issuecomment-2783895679

Let's move the discussion there.